packages/eslint-plugin/src/rules/require-types-exports.ts

```diff
@@ -0,0 +1,241 @@
+      node: TSESTree.ExportNamedDeclaration & {
+        declaration: TSESTree.VariableDeclaration;
+      },
+    ): void {
+      for (const declaration of node.declaration.declarations) {
+        if (declaration.init?.type === AST_NODE_TYPES.ArrowFunctionExpression) {
+          checkFunctionTypes(declaration.init);
+        } else {
+          checkVariableTypes(declaration);
+        }
+      }
+    }
+
+    function visitDefaultExportedArrowFunction(
+      node: TSESTree.ExportDefaultDeclaration & {
+        declaration: TSESTree.ArrowFunctionExpression;
+      },
+    ): void {
+      checkFunctionTypes(node.declaration);
+    }
+
+    function visitDefaultExportedIdentifier(
+      node: TSESTree.DefaultExportDeclarations & {
+        declaration: TSESTree.Identifier;
+      },
+    ): void {
+      const scope = context.sourceCode.getScope(node);
+      const variable = scope.set.get(node.declaration.name);
+
+      if (!variable) {
+        return;
+      }
+
+      for (const def of variable.defs) {
+        if (def.type !== DefinitionType.Variable || !def.node.init) {
+          continue;
+        }
+
+        if (
+          def.node.init.type === AST_NODE_TYPES.ArrowFunctionExpression ||
+          def.node.init.type === AST_NODE_TYPES.FunctionExpression
+        ) {
+          checkFunctionTypes(def.node.init);
+        } else {
+          checkVariableTypes(def.node);
+        }
+      }
+    }
+
+    function checkFunctionTypes(node: FunctionNode): void {
+      const scope = context.sourceCode.getScope(node);
+
+      scope.through
+        .map(ref => ref.identifier.parent)
+        .filter(
+          (node): node is TSESTree.TSTypeReference =>
+            node.type === AST_NODE_TYPES.TSTypeReference,
+        )
+        .forEach(checkTypeNode);
+    }
+
+    function checkVariableTypes(
+      node: TSESTree.LetOrConstOrVarDeclarator,
+    ): void {
+      typeReferences.forEach(r => {
+        // TODO: Probably not the best way to do it...
+        if (isLocationOverlapping(r.identifier.loc, node.loc)) {
+          checkTypeNode(r.identifier.parent);
+        }
+      });
+    }
+
+    function checkTypeNode(node: TSESTree.TSTypeReference): void {
+      const name = getTypeName(node);
+
+      if (!name) {
+        // TODO: Report the whole function? Is this case even possible?
+        return;
+      }
+
+      const isExternalized = externalizedTypes.has(name);
+      const isReported = reportedTypes.has(name);
+
+      if (isExternalized || isReported) {
+        return;
+      }
+
+      context.report({
+        node: node,
+        messageId: 'requireTypeExport',
+        data: {
+          name,
+        },
+      });
+
+      reportedTypes.add(name);
+    }
+
+    function getTypeName(typeReference: TSESTree.TSTypeReference): string {
+      if (typeReference.typeName.type === AST_NODE_TYPES.Identifier) {
+        return typeReference.typeName.name;
+      }
+
+      return '';
```

Comment:
From the ast spec:

```ts
export declare type EntityName = Identifier | ThisExpression | TSQualifiedName;
```

This'll need to handle `this` and namespace types, no?

```ts
export class Wrapper {
  work(other: this) {}
}
```

Reply:
Yeah... IDK why I haven't looked at spec before writing this 😅 
Thanks!